staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,750 @@
+			klog.V(3).Infof("Retaining mandatory priority level %q despite lack of API object", plName)
+		} else {
+			if plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle() {
+				// Either there are no queues or they are done
+				// draining and no use is coming from another
+				// goroutine
+				klog.V(3).Infof("Removing undesired priority level %q, Type=%v", plName, plState.config.Type)
+				continue
+			}
+			if !plState.quiescing {
+				klog.V(3).Infof("Priority level %q became undesired", plName)
+				plState.quiescing = true
+			}
+		}
+		var err error
+		plState.qsCompleter, err = qscOfPL(meal.cfgCtl.queueSetFactory, plState.queues, plName, &plState.config, meal.cfgCtl.requestWaitLimit)
+		if err != nil {
+			// This can not happen because qscOfPL already approved this config
+			panic(fmt.Sprintf("%s from name=%q spec=%#+v", err.Error(), plName, fcfmt.Fmt(plState.config)))
+		}
+		if plState.config.Limited != nil {
+			meal.shareSum += float64(plState.config.Limited.AssuredConcurrencyShares)
+		}
+		meal.haveExemptPL = meal.haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameExempt
+		meal.haveCatchAllPL = meal.haveCatchAllPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll
+		meal.newPLStates[plName] = plState
+	}
+}
+
+// For all the priority levels of the new config, divide up the
+// server's total concurrency limit among them and create/update their
+// QueueSets.
+func (meal *cfgMeal) finishQueueSetReconfigsLocked() {
+	for plName, plState := range meal.newPLStates {
+		if plState.config.Limited == nil {
+			klog.V(5).Infof("Using exempt priority level %q: quiescing=%v", plName, plState.quiescing)
+			continue
+		}
+
+		// The use of math.Ceil here means that the results might sum
+		// to a little more than serverConcurrencyLimit but the
+		// difference will be negligible.
+		concurrencyLimit := int(math.Ceil(float64(meal.cfgCtl.serverConcurrencyLimit) * float64(plState.config.Limited.AssuredConcurrencyShares) / meal.shareSum))
+		metrics.UpdateSharedConcurrencyLimit(plName, concurrencyLimit)
+
+		if plState.queues == nil {
+			klog.V(5).Infof("Introducing queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescing=%v (shares=%v, shareSum=%v)", plName, fcfmt.Fmt(plState.config), concurrencyLimit, plState.quiescing, plState.config.Limited.AssuredConcurrencyShares, meal.shareSum)
+		} else {
+			klog.V(5).Infof("Retaining queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescing=%v, numPending=%d (shares=%v, shareSum=%v)", plName, fcfmt.Fmt(plState.config), concurrencyLimit, plState.quiescing, plState.numPending, plState.config.Limited.AssuredConcurrencyShares, meal.shareSum)
+		}
+		plState.queues = plState.qsCompleter.Complete(fq.DispatchingConfig{ConcurrencyLimit: concurrencyLimit})
```

Comment:
This is discussed in the KEP, see https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#reaction-to-configuration-changes .

Now that you raise the question, I see it is not clearly documented in the QueueSet interface.  I will update that.

Reply:
Ah, okay, I think what's happening here is that because any already-queued Requests have pointers to the QueueSet on which they were created (rather than indirecting through their priority level), we can safely drop plState.queues and it won't get GCed until after its queues actually empty out.

In that case, though, why do we bother with the complicated machinery for quiescing plStates? The same logic ought to apply, so long as we make sure to create and enqueue requests from Match().

Reply:
Note that when starting from a particular QueueSet X, calling X.BeginConfigChange(), taking the QueueSetCompleter returned from that, then calling its Complete method is defined to return the original QueueSet X (with side effects having happened).  So the line being discussed here is not throwing anything away.  Quiescing is about retaining a priority level, including its QueueSet, while its queues drain after its corresponding PriorityLevelConfiguration object is deleted.

Reply:
Updated QueueSet commentary to emphasize this.